File: apis/ocpdiag/ocpdiag/core/hwinterface/lib/off_dut_machine_interface/ssh/remote_test.cc
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "ocpdiag/core/hwinterface/lib/off_dut_machine_interface/ssh/remote.h"

#include <csignal>
#include <cstdio>
#include <filesystem>
#include <fstream>
#include <memory>
#include <sstream>
#include <string_view>
#include <utility>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "absl/time/time.h"
#include "absl/types/span.h"
#include "ocpdiag/core/hwinterface/lib/off_dut_machine_interface/remote.h"
#include "ocpdiag/core/testing/status_matchers.h"

namespace ocpdiag::hwinterface {
namespace remote {

using ::ocpdiag::testing::StatusIs;
using ::testing::Eq;

// constexpr absl::string_view kSshPath = "/usr/bin/ssh";

// SshTest is for testing the SSHConnInterface. It creates an executable script
// named "ssh" in tmp folder. By modifying PATH, we can let execvp to execute
// the script as a fake ssh. Then we can put different command in the script to
// veirfy the SSHConnInterface behavior.
class SshTest : public ::testing::Test {
 public:
  SshTest() : SshTest("", "") {}
  SshTest(absl::string_view ssh_key_path,
          absl::string_view ssh_tunnel_file_path)
      : ssh_conn_(NodeSpec{"dut"}, ssh_key_path, ssh_tunnel_file_path,
                  ssh_path_.string()) {}

 protected:
  static void SetUpTestSuite() {
    // Update the path to prepend the temporary directory.
    std::filesystem::path tmpdir = std::filesystem::temp_directory_path();
    absl::string_view old_path = getenv("PATH");
    // PATH should be appended with colon.
    std::string new_path = absl::StrCat(tmpdir.native(), ":", old_path);
    setenv("PATH", new_path.c_str(), /*__replace=*/true);
    // Create an executable SSH script.
    ssh_path_ = tmpdir / "ssh";
    std::ofstream(ssh_path_, std::ios_base::out | std::ios_base::trunc)
        << "#!/bin/sh" << std::endl;
    std::filesystem::permissions(ssh_path_, std::filesystem::perms::all);
  }

  void SetUp() {
    // Reset SSH behavior on each test.
    std::ofstream(ssh_path_, std::ios_base::out | std::ios_base::trunc)
        << "#!/bin/sh" << std::endl;
  }

  void FakeSsh(absl::string_view script) {
    std::ofstream(ssh_path_, std::ios_base::out | std::ios_base::trunc)
        << "#!/bin/sh" << std::endl
        << script << std::endl;
  }

  absl::StatusOr<ConnInterface::CommandResult> RunCommand(
      absl::string_view command,
      ConnInterface::CommandOption options = ConnInterface::CommandOption()) {
    absl::Span<absl::string_view> commands = absl::MakeSpan(&command, 1);
    return ssh_conn_.RunCommand(absl::Seconds(1), commands, options);
  }

  absl::StatusOr<absl::Cord> ReadFile(absl::string_view file_name) {
    return ssh_conn_.ReadFile(file_name);
  }

  absl::Status WriteFile(absl::string_view file_name, absl::string_view data) {
    return ssh_conn_.WriteFile(file_name, data);
  }

  std::string ReadLocalFile(std::string_view file_name) {
    std::stringstream buffer;
    buffer << std::ifstream(std::string(file_name).c_str()).rdbuf();
    return buffer.str();
  }

  static std::filesystem::path ssh_path_;
  SSHConnInterface ssh_conn_;
};  // class SshTest

std::filesystem::path SshTest::ssh_path_;

TEST_F(SshTest, SshStdoutCaptured) {
  FakeSsh("echo -n 'Hello World!'");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->stdout, Eq("Hello World!"));
}

TEST_F(SshTest, SshStderrCaptured) {
  FakeSsh(">&2 echo -n 'Hello World!'");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->stderr, Eq("Hello World!"));
}

TEST_F(SshTest, SshUserHostExpected) {
  FakeSsh("echo -n $1");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->exit_code, Eq(0));
  EXPECT_THAT(result->stdout, Eq("root@dut"));
}

TEST_F(SshTest, SshExitCodeCaptured) {
  FakeSsh("exit 42");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->exit_code, Eq(42));
}

TEST_F(SshTest, SshSignalCaptured) {
  FakeSsh("kill -9 $$");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->exit_code, Eq(-9));
}

TEST_F(SshTest, SshDeadlineExceeded) {
  FakeSsh("sleep 100");
  auto result = RunCommand("ls");
  EXPECT_THAT(result, StatusIs(absl::StatusCode::kDeadlineExceeded));
}

TEST_F(SshTest, SshStdoutRedirectRejected) {
  auto result =
      RunCommand("ls", ConnInterface::CommandOption{"stdout_file", ""});
  EXPECT_THAT(result, StatusIs(absl::StatusCode::kUnimplemented));
}

TEST_F(SshTest, SshStderrRedirectRejected) {
  auto result =
      RunCommand("ls", ConnInterface::CommandOption{"", "stderr_file"});
  EXPECT_THAT(result, StatusIs(absl::StatusCode::kUnimplemented));
}

TEST_F(SshTest, SshReadFileCallCat) {
  FakeSsh("echo $2; echo -n $3");
  auto result = ReadFile("file_to_read");
  ASSERT_OK(result);
  EXPECT_THAT(*result, Eq(absl::Cord("cat\nfile_to_read")));
}

TEST_F(SshTest, SshReadFileReturnStdoutAsContent) {
  FakeSsh("echo -n \"abc\x01\x05\x0a\x15\"");
  auto result = ReadFile("file_to_read");
  ASSERT_OK(result);
  EXPECT_THAT(*result, Eq(absl::Cord("abc\x01\x05\x0a\x15")));
}

TEST_F(SshTest, SshWriteFileCallCat) {
  // Tmp file to get the stdout of fakessh.
  std::string tmp_file = std::tmpnam(nullptr);
  FakeSsh("echo -n \"$2\n$3\n$4\" >> " + tmp_file);

  auto result = WriteFile("file_to_write", "");

  ASSERT_OK(result);
  std::string actual = ReadLocalFile(tmp_file);
  EXPECT_THAT(actual, Eq("cat\n>\nfile_to_write"));
}

TEST_F(SshTest, SshWriteFileFeedDataToStdin) {
  // Tmp file to get the stdout of fakessh.
  auto tmp_file = std::tmpnam(nullptr);
  FakeSsh(absl::StrCat("cat >> ", tmp_file));

  auto result = WriteFile("file_to_write", "abc\x01\x05\x0a\x15");

  ASSERT_OK(result);
  std::string actual = ReadLocalFile(tmp_file);
  EXPECT_THAT(actual, Eq("abc\x01\x05\x0a\x15"));
}

class SshWithKeyPathTest : public SshTest {
 public:
  SshWithKeyPathTest() : SshTest("ssh_key_path", "") {}
};

TEST_F(SshWithKeyPathTest, SshUserHostExpected) {
  FakeSsh("echo -n $2 $3");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->exit_code, Eq(0));
  EXPECT_THAT(result->stdout, Eq("-i ssh_key_path"));
}

class SshWithTunnelFilePathTest : public SshTest {
 public:
  SshWithTunnelFilePathTest() : SshTest("", "ssh_tunnel_file_path") {}
};

TEST_F(SshWithTunnelFilePathTest, SshUserHostExpected) {
  FakeSsh("echo -n $2 $3");
  auto result = RunCommand("ls");
  ASSERT_OK(result);
  EXPECT_THAT(result->exit_code, Eq(0));
  EXPECT_THAT(result->stdout, Eq("-S ssh_tunnel_file_path"));
}

}  // namespace remote
}  // namespace ocpdiag::hwinterface

File: apis/ocpdiag/ocpdiag/core/hwinterface/lib/off_dut_machine_interface/ssh/remote_test.py
# Copyright 2022 Google LLC
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

"""Tests for ocpdiag.core.hwinterface.lib.off_dut_machine_interface.ssh.remote."""

import datetime
import signal
import subprocess
from unittest import mock

import unittest
from absl.testing import parameterized
from ocpdiag.core.hwinterface.lib.off_dut_machine_interface import remote
from ocpdiag.core.hwinterface.lib.off_dut_machine_interface.ssh import remote as ssh_remote


class RemoteTest(parameterized.TestCase):

  def setUp(self):
    super().setUp()
    self.addCleanup(mock.patch.stopall)
    self.run_fn = mock.patch.object(subprocess, 'run', autospec=True).start()
    self.run_fn.return_value = subprocess.CompletedProcess(
        args='',
        returncode=0,
        stdout=bytes([0x1, 0xaa, 0xff]),
        stderr='stderr'.encode())
    self.conn = ssh_remote.SSHConn(
        remote.NodeSpec(address='dut'),
        ssh_key_path='',
        ssh_tunnel_file_path='',
        ssh_bin_path='/bin/ssh')

  def test_runcommand_calls_subprocess_run(self):
    result = self.conn.RunCommand(
        duration=datetime.timedelta(minutes=15),
        stdin=bytes(),
        args=['ls', '-l'])

    # Verify the result.
    self.assertIsNotNone(result)
    self.assertEqual(0, result.exit_code)
    self.assertEqual(bytes([0x1, 0xaa, 0xff]), result.stdout)
    self.assertEqual('stderr'.encode(), result.stderr)

    # Verify correct argument to subprocess.run.
    self.run_fn.assert_called_once_with(
        args=['/bin/ssh', 'root@dut', 'ls', '-l'],
        input=bytes(),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=900,
        check=False)

  def test_readfile_calls_subprocess_run(self):
    result = self.conn.ReadFile('/to_read')

    # Verify the result.
    self.assertEqual(bytes([0x1, 0xaa, 0xff]), result)

    # Verify correct argument to subprocess.run.
    self.run_fn.assert_called_once_with(
        args=['/bin/ssh', 'root@dut', 'cat', '/to_read'],
        input=bytes(),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=900,
        check=False)

  def test_writefile_calls_subprocess_run(self):
    self.conn.WriteFile(path='/to_write', data=bytes([0x3, 0xff, 0x1d]))

    # Verify correct argument to subprocess.run.
    self.run_fn.assert_called_once_with(
        args=['/bin/ssh', 'root@dut', 'cat', '>', '/to_write'],
        input=bytes([0x3, 0xff, 0x1d]),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=900,
        check=False)

  def test_ssh_key_path_is_used(self):
    self.conn = ssh_remote.SSHConn(
        remote.NodeSpec(address='dut'),
        ssh_key_path='the_ssh_key_path',
        ssh_tunnel_file_path='',
        ssh_bin_path='/bin/ssh')
    self.conn.ReadFile(path='/to_read')

    # Verify correct argument to subprocess.run.
    self.run_fn.assert_called_once_with(
        args=[
            '/bin/ssh', 'root@dut', '-i', 'the_ssh_key_path', 'cat', '/to_read'
        ],
        input=bytes(),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=900,
        check=False)

  def test_ssh_tunnel_file_path_is_used(self):
    self.conn = ssh_remote.SSHConn(
        remote.NodeSpec(address='dut'),
        ssh_key_path='',
        ssh_tunnel_file_path='the_ssh_tunnel_file_path',
        ssh_bin_path='/bin/ssh')
    self.conn.ReadFile(path='/to_read')

    # Verify correct argument to subprocess.run.
    self.run_fn.assert_called_once_with(
        args=[
            '/bin/ssh', 'root@dut', '-S', 'the_ssh_tunnel_file_path', 'cat',
            '/to_read'
        ],
        input=bytes(),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        timeout=900,
        check=False)

  def test_readfile_raise_when_read_fail(self):
    self.run_fn.return_value = subprocess.CompletedProcess(
        args='', returncode=-signal.SIGKILL, stdout=bytes(), stderr=bytes())
    with self.assertRaises(IOError):
      self.conn.ReadFile(path='/to_read')

  def test_readfile_raise_when_write_fail(self):
    self.run_fn.return_value = subprocess.CompletedProcess(
        args='', returncode=-signal.SIGKILL, stdout=bytes(), stderr=bytes())
    with self.assertRaises(IOError):
      self.conn.WriteFile(path='/to_write', data=bytes([0x1]))

  @parameterized.named_parameters(('redirect_stdout', 'std_out', ''),
                                  ('redirect_stderr', '', 'std_err'))
  def test_runcommand_throws_for_redirect(self, stdout_file, stderr_file):
    with self.assertRaises(NotImplementedError):
      self.conn.RunCommand(
          duration=datetime.timedelta(minutes=15),
          stdin=bytes(),
          args=['ls', '-l'],
          opts=remote.CommandOption(
              stdout_file=stdout_file, stderr_file=stderr_file))

  def test_runcommand_throws_when_timeout(self):
    self.run_fn.side_effect = subprocess.TimeoutExpired(cmd='ls -l', timeout=1)
    with self.assertRaises(TimeoutError):
      self.conn.RunCommand(
          duration=datetime.timedelta(seconds=1),
          stdin=bytes(),
          args=['ls', '-l'])

  def test_runcommand_throws_when_fail(self):
    self.run_fn.side_effect = subprocess.SubprocessError()
    with self.assertRaises(RuntimeError):
      self.conn.RunCommand(
          duration=datetime.timedelta(seconds=1),
          stdin=bytes(),
          args=['ls', '-l'])


if __name__ == '__main__':
  unittest.main()
